arbnode/dataposter/data_poster.go

```diff
@@ -478,106 +516,158 @@ func (p *DataPoster) feeAndTipCaps(ctx context.Context, nonce uint64, gasLimit u
 			} else {
-				balanceForTx.Mul(balanceForTx, common.Big2)
+				balancePerWeight.Mul(balancePerWeight, common.Big2)
 			}
-			balanceForTx.Div(balanceForTx, common.Big3)
+			balancePerWeight.Div(balancePerWeight, common.Big3)
 			// After weighting, split the balance between each of the transactions
 			// other than the first tx which already got half.
 			// balanceForTx /= config.MaxMempoolTransactions-1
-			balanceForTx.Div(balanceForTx, arbmath.UintToBig(config.MaxMempoolTransactions-1))
+			balancePerWeight.Div(balancePerWeight, arbmath.UintToBig(maxMempoolWeight-1))
+			balanceForTx.Add(balanceForTx, arbmath.BigMulByUint(balancePerWeight, weight))
 		}
 	}
-	// TODO: take into account blob costs
-	balanceFeeCap := arbmath.BigDivByUint(balanceForTx, gasLimit)
-	if arbmath.BigGreaterThan(newFeeCap, balanceFeeCap) {
+
+	if arbmath.BigGreaterThan(targetMaxCost, balanceForTx) {
 		log.Warn(
 			"lack of L1 balance prevents posting transaction with desired fee cap",
 			"balance", latestBalance,
-			"maxTransactions", config.MaxMempoolTransactions,
+			"weight", weight,
+			"maxMempoolWeight", maxMempoolWeight,
 			"balanceForTransaction", balanceForTx,
 			"gasLimit", gasLimit,
-			"desiredFeeCap", newFeeCap,
-			"balanceFeeCap", balanceFeeCap,
+			"targetMaxCost", targetMaxCost,
 			"nonce", nonce,
 			"softConfNonce", softConfNonce,
 		)
-		newFeeCap = balanceFeeCap
-	}
-
-	if arbmath.BigGreaterThan(newTipCap, newFeeCap) {
-		log.Warn(
-			"reducing new tip cap to new fee cap",
+		targetMaxCost = balanceForTx
+	}
+
+	// Divide the targetMaxCost into blob and non-blob costs.
+	currentNonBlobFee := arbmath.BigAdd(latestHeader.BaseFee, newTipCap)
+	blobGasUsed := params.BlobTxBlobGasPerBlob * numBlobs
+	currentBlobCost := arbmath.BigMulByUint(currentBlobFee, blobGasUsed)
+	currentNonBlobCost := arbmath.BigMulByUint(currentNonBlobFee, gasLimit)
+	newBlobFeeCap := arbmath.BigMul(targetMaxCost, currentBlobFee)
+	newBlobFeeCap.Div(newBlobFeeCap, arbmath.BigAdd(currentBlobCost, currentNonBlobCost))
+	if lastTx != nil && lastTx.BlobGasFeeCap() != nil {
+		newBlobFeeCap = arbmath.BigMax(newBlobFeeCap, arbmath.BigMulByBips(lastTx.BlobGasFeeCap(), minBlobRbfIncrease))
+	}
+	targetBlobCost := arbmath.BigMulByUint(newBlobFeeCap, blobGasUsed)
+	targetNonBlobCost := arbmath.BigSub(targetMaxCost, targetBlobCost)
+	newBaseFeeCap := arbmath.BigDivByUint(targetNonBlobCost, gasLimit)
+	if lastTx != nil && numBlobs > 0 && arbmath.BigDivToBips(newBaseFeeCap, lastTx.GasFeeCap()) < minRbfIncrease {
+		// Reduce the blobfee cap to ensure that the basefee meats the minimum rbf increase
+		newBaseFeeCap = arbmath.BigMulByBips(lastTx.GasFeeCap(), minRbfIncrease)
+		newNonBlobCost := arbmath.BigMulByUint(newBaseFeeCap, gasLimit)
+		baseFeeCostIncrease := arbmath.BigSub(targetBlobCost, newNonBlobCost)
+		newBlobCost := arbmath.BigSub(targetBlobCost, baseFeeCostIncrease)
+		newBlobFeeCap = arbmath.BigDivByUint(newBlobCost, blobGasUsed)
```

Comment:
The `newBlobFeeCap` could now no longer meet the `minRbfIncrease` constraint at this point, right? Is there anything we should do in that case?

Reply:
Great question. In this case, we simply don't have enough max cost allocated to this tx yet, so we can't fulfill the rbf rules with either our current balance or our fee escalation formula. replaceTx will see that the returned fees are under the min rbf threshold and not attempt to rbf. As time elapses, the fee escalation formula should continue recommending a higher and higher fee, and we'll eventually clear the threshold. And if it's low on balance we'll send it more eth and it'll be able to continue.